delfin/drivers/dell_emc/power_store/rest_handler.py

```diff
@@ -0,0 +1,958 @@
+        elif res.status_code == consts.StatusCode.PARTIAL_CONTENT:
+            result.extend(res.json())
+            if len(res.json()) == consts.DigitalConstant.TWO_THOUSAND_INT:
+                offset = offset + consts.DigitalConstant.TWO_THOUSAND_INT
+                self.rest_call(url, data, method, offset, result, count)
+        elif res.status_code == consts.StatusCode.UNAUTHORIZED or \
+                res.status_code == consts.StatusCode.FORBIDDEN:
+            if count < consts.DigitalConstant.THREE_INT:
+                self.login()
+                count = count + consts.DigitalConstant.ONE_INT
+                self.rest_call(url, data, method, offset, result, count)
+        return result
+
+    def get_storage(self, storage_id):
+        clusters = self.rest_call(self.REST_CLUSTER_URL)
+        if not clusters:
+            LOG.error('The cluster data is empty')
+            raise exception.StorageBackendException(
+                'The cluster data is empty')
+        cluster = clusters[consts.DigitalConstant.ZERO_INT]
+        appliance_id = cluster.get('primary_appliance_id')
+        appliances = self.rest_call(self.REST_APPLIANCE_URL)
+        model = ''
+        for appliance in appliances:
+            if appliance_id == appliance.get('id'):
+                model = appliance.get('model')
+        pools = self.get_storage_pools(storage_id)
+        total_capacity = consts.DigitalConstant.ZERO_INT
+        used_capacity = consts.DigitalConstant.ZERO_INT
+        for pool in pools:
+            total_capacity += pool.get('total_capacity')
+            used_capacity += pool.get('used_capacity')
+        disks = self.get_disks(storage_id)
+        raw_capacity = consts.DigitalConstant.ZERO_INT
+        for disk in disks:
+            raw_capacity += disk.get('capacity')
+        storage_result = {
+            'model': model,
+            'total_capacity': total_capacity,
+            'raw_capacity': raw_capacity,
+            'used_capacity': used_capacity,
+            'free_capacity': total_capacity - used_capacity,
+            'vendor': 'DELL EMC',
+            'name': cluster.get('name'),
+            'serial_number': cluster.get('global_id'),
+            'firmware_version': self.get_firmware_version(appliance_id),
+            'status': consts.STORAGE_STATUS_MAP.get(
+                cluster.get('state'), constants.StorageStatus.UNKNOWN)
+        }
+        return storage_result
+
+    def get_firmware_version(self, appliance_id):
+        firmware_version = ''
+        software_s = self.rest_call(RestHandler.REST_SOFTWARE_INSTALLED_URL)
+        for software in software_s:
+            appliance_d = software.get('appliance')
+            if not appliance_d:
+                continue
+            software_appliance_id = appliance_d.get('id')
+            if appliance_id == software_appliance_id:
+                firmware_version = software.get('release_version')
+                break
```

Comment:
recommended type like:
```
if appliance_id == appliance_d.get('id'):
    return software.get('release_version')
```

Reply:
It has been modified.

return software.get('release_version')